src/main.cpp

```diff
@@ -1,6 +1,223 @@
+		// 				client_count++;
+		// 			}
+		// 		}
+		// 		else
+		// 		{
+		// 			//check request timeout before recv and set the bool of timeout to false and continue???
+		// 			//recv from client and add to request buffer
+		// 			ssize_t recv_len = sm.recv_append(pfds[i].fd, recv_buf);
+		// 			//1st timestamp for timeout, using Maybe<time_t> init_time;
+		// 			if (recv_len <= 0)
+		// 			{
+		// 				close(pfds[i].fd);
+		// 				pollfds::delete_client_fd(pfds, i);
+		// 				client_count--;
+		// 			}
+		// 			else
+		// 			{
+		// 				//parse request
+		// 				//process request
+		// 			}
+		// 		}
+		// 	}
+		// 	else if (pfds[i].revents & POLLOUT)
+		// 	{
+		// 		//send response to client
+		// 		ssize_t send_len = sm.send_all(pfds[i].fd);
+		// 		//TODO: do I need to close the client socket if send fails?
+		// 		if (send_len == -1)
+		// 		{
+		// 			close(pfds[i].fd);
+		// 			pollfds::delete_client_fd(pfds, i);
+		// 			client_count--;
+		// 		}
+		// 	}
+		// }
+
+		//check events for server sockets
+		for (int i = 0; i < server_socket_count; i++)
+		{
+			if (pfds[i].revents & POLLIN)
+			{
+				if (client_count < max_clients)
+				{
+					//accept connection if max clients not reached
+					int client_socket = sm.accept_client(pfds[i].fd);
+					if (client_socket == -1)
+						continue ;
+					else
+					{
+						//add client to pollfd
+						pollfds::add_client_fd(pfds, client_socket);
+						client_count++;
+					}
+				}
+			}
+		}
+		//check events for client sockets
+		for (int i = server_socket_count; i < pfds.size(); i++)
+		{
+			//check if client socket is timeout
+			bool timeout = sm.is_timeout(pfds[i].fd);
+			if ((pfds[i].revents & POLLIN) == 0 && (pfds[i].revents & POLLOUT) == 0) //no client events
+			{
+				if (timeout == true)
+				{
+					close(pfds[i].fd);
+					pollfds::delete_client_fd(pfds, i);
+					client_count--;
+				}
+				continue;
+			}
+			//socket is ready for reading
+			if (pfds[i].revents & POLLIN)
+			{
+				if (timeout == true)
+				{
+					//generate 408 Request Timeout and build std::string response in client
+					pfds[i].events = POLLOUT;
+					continue;
+				}
+				//recv from client and add to request buffer
+				ssize_t recv_len = sm.recv_append(pfds[i].fd, recv_buf);
+				//1st timestamp for timeout, using Maybe<time_t> first_recv_time in the else block of recv_append()
+				if (recv_len <= 0)
+				{
+					close(pfds[i].fd);
+					pollfds::delete_client_fd(pfds, i);
+					client_count--;
+				}
+				else
+				{
+					sm.set_first_recv_time(pfds[i].fd); //set first recv time if it is not set
+					//if request exceeds max body size
+						//set pfds[i].events = POLLOUT;
+						//generate 413 Request Entity Too Large and build std::string response in client
+					//if request is complete
+						//set pfds[i].events = POLLOUT;
+						//parse request
+						//process request
+				}
+			}
+			//socket is ready for writing
+			else if (pfds[i].revents & POLLOUT)
```

Comment:
probably not `else if`, but just `if`. It is rarely the case, but imagine a client (hacker) trying to flood the server by sending a lot of data, so this socket is always POLLIN. But because it is a `else if`, you will never send things to the client, so the string that you accumulate will just get bigger and bigger until it does not have enough memory.